src/hotspot/share/c1/c1_ValueMap.cpp

```diff
@@ -394,7 +416,11 @@ void LoopInvariantCodeMotion::process_block(BlockBegin* block) {
       cur_invariant = is_invariant(cvt->value());
     }
 
-    if (cur_invariant) {
+    // Check that insertion point has higher dom depth than all inputs to cur
+    CheckInsertionPoint v(_insertion_point);
+    cur->input_values_do(&v);
```

Comment:
For compilation efficiency, would it be possible to perform this computation only when `cur_invariant` holds? You could for example encapsulate the creation of `v`, call to `cur->input_values_do(&v)`, and `v.is_valid()` check into an auxiliary function `is_dominated_by_inputs(_insertion_point, cur)` or similar and call that function below (`if (cur_invariant && is_dominated_by_inputs(_insertion_point, cur)) {`).